docs/proposals/13-automated-seed-management.md

```diff
@@ -0,0 +1,369 @@
+
+The primary metric used for scaling should be the number of shoots already scheduled onto that seed either as a direct value or as a percentage of the seed's capacity for shoots introduced in [Ensuring Seeds Capacity for Shoots Is Not Exceeded](#ensuring-seeds-capacity-for-shoots-is-not-exceeded) (*utilization*). Later, custom metrics based on other resources, including provider-specific resources, could be considered as well.
+
+**Note:** Even if the controller is called *Horizontal Pod Autoscaler*, it is capable of scaling any resource with a `scale` subresource, using any custom metric. Therefore, initially it was proposed to use this controller directly. However, a number of important drawbacks were identified with this approach, and so it is no longer proposed here.
+
+### SeedAutoscaler
+
+The SeedAutoscaler automatically scales the number of [ShootedSeeds](#shootedseeds) in a [ShootedSeedSet](#shootedseedsets) based on observed resource utilization. The resource could be any resource that is tracked via the `capacity` and `allocatable` fields in the Seed status, including in particular the number of shoots already scheduled onto the seed.
+
+#### SeedAutoscaler Resource
+
+The SeedAutoscaler is implemented as a custom resource and a new controller. The resource determines the behavior of the controller. The `SeedAutoscaler` resource has a `scaleTargetRef` that specifies the target resource to be scaled, the minimum and maximum number of replicas, as well as a list of metrics. The only supported metric type initially is `Resource` for resources that are tracked via the `capacity` and `allocatable` fields in the Seed status. The resource target can be of type `Utilization` or `AverageValue`.
+
+Example `SeedAutoscaler` resource:
+
+```yaml
+apiVersion: seedmanagement.gardener.cloud/v1alpha1
+kind: SeedAutoscaler
+metadata:
+  name: crazy-botany
+  namespace: garden
+spec:
+  scaleTargetRef:
+    apiVersion: seedmanagement.gardener.cloud/v1alpha1
+    kind: ShootedSeedSet
+    name: crazy-botany
+  minReplicas: 1
+  maxReplicas: 10
+  metrics:
+  - type: Resource # Only Resource is supported
+    resource:
+      name: shoots
+      target:
+        type: Utilization # Utilization or AverageValue
+        averageUtilization: 50
+```
+
+#### SeedAutoscaler Controller
+
+`SeedAutoscaler` resources are reconciled by a new controller, either in `gardenlet` or out-of-tree, similarly to [cluster-autoscaler](https://github.com/gardener/autoscaler). The controller periodically adjusts the number of replicas in a ShootedSeedSet to match the observed average resource utilization to the target specified by user.
```

Comment:
Can you explain how the `.spec.metrics` are evaluated, i.e., would the controller look at the `Seed`'s `.status.capacity`?

Reply:
OK, I will describe this in somewhat more detail.

Reply:
I ended up describing two different approaches but couldn't decide which one is best:

> The metrics the the controller uses, for example the `shoots` metric in the above example, could be evaluated in one of the following ways:
>
> * Directly, by looking at the `capacity` and `allocatable` fields in the Seed status and comparing to the actual resource consumption calculated by simply counting all shoots that meet a certain criteria (e.g. shoots that are scheduled onto the seed), then taking an average over all seeds in the set.
> * By sampling existing metrics exported for example by [`gardener-metrics-exporter`](https://github.com/gardener/gardener-metrics-exporter). 
>
> The second approach decouples the seed autoscaler controller from the actual metrics evaluation, and therefore allows plugging in new metrics more easily. It also has the advantage that the exported metrics could also be used for other purposes, e.g. for triggering Prometheus alerts or building Grafana dashboards. It has the disadvantage that the seed autoscaler controller would depend on the metrics exporter to do its job properly. 

Do you have a strong opinion which of the two options is to be preferred? If not, I guess we could leave them both in the proposal and discuss them again later.